File: src/main/java/komu/blunt/analyzer/PatternAnalyzer.java
package komu.blunt.analyzer;

import komu.blunt.core.*;
import komu.blunt.types.patterns.*;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

final class PatternAnalyzer {

    public CoreExpression createExtractor(Pattern pattern, VariableReference matchedObject, StaticEnvironment env) {
        return pattern.accept(new ExtractorCreator(matchedObject, env), PatternPath.EMPTY);
    }

    private static final class ExtractorCreator implements PatternVisitor<PatternPath,CoreExpression> {
        private final VariableReference matchedObject;
        private final StaticEnvironment env;

        ExtractorCreator(VariableReference matchedObject, StaticEnvironment env) {
            this.matchedObject = checkNotNull(matchedObject);
            this.env = checkNotNull(env);
        }

        @Override
        public CoreExpression visit(WildcardPattern pattern, PatternPath path) {
            return new CoreConstantExpression(true);
        }

        @Override
        public CoreExpression visit(VariablePattern pattern, PatternPath path) {
            VariableReference var = env.lookupInCurrentScopeOrDefine(pattern.var);

            return new CoreSequenceExpression(
                new CoreSetExpression(var, new CoreExtractExpression(matchedObject, path)),
                new CoreConstantExpression(true));
        }

        @Override
        public CoreExpression visit(ConstructorPattern pattern, PatternPath path) {
            List<CoreExpression> exps = new ArrayList<>(1 + pattern.args.size());
            exps.add(matchesConstructor(path, pattern.name));

            int i = 0;
            for (Pattern p : pattern.args) 
                exps.add(p.accept(this, path.extend(i++)));

            return CoreExpression.and(exps);
        }

        private CoreExpression matchesConstructor(PatternPath path, String name) {
            return new CoreEqualConstantExpression(name, new CoreExtractTagExpression(matchedObject, path));
        }

        @Override
        public CoreExpression visit(LiteralPattern pattern, PatternPath path) {
            return new CoreEqualConstantExpression(pattern.value, new CoreExtractExpression(matchedObject, path));
        }
    }
}

File: src/main/kotlin/komu/blunt/analyzer/PatternAnalyzer.kt
package komu.blunt.analyzer

import komu.blunt.core.*
import komu.blunt.types.patterns.Pattern
import java.util.*

object PatternAnalyzer {

    fun makeExtractor(pattern: Pattern, matchedObject: VariableReference, env: StaticEnvironment, path: PatternPath=PatternPath.EMPTY): CoreExpression = when (pattern) {
        is Pattern.Wildcard     -> CoreExpression.EMPTY
        is Pattern.Literal      -> CoreExpression.EMPTY
        is Pattern.Variable     -> variableExtractor(pattern, path, env, matchedObject)
        is Pattern.Constructor  -> constructorExtractor(pattern, path, env, matchedObject)
    }

    private fun variableExtractor(pattern: Pattern.Variable, path: PatternPath, env: StaticEnvironment, matchedObject: VariableReference): CoreExpression {
        val v = env.lookupInCurrentScopeOrDefine(pattern.variable)
        return CoreSetExpression(v, CoreExtractExpression(matchedObject, path))
    }

    private fun constructorExtractor(pattern: Pattern.Constructor, path: PatternPath, env: StaticEnvironment, matchedObject: VariableReference): CoreExpression =
        CoreExpression.sequence(pattern.args.mapIndexed { i, p ->
            makeExtractor(p, matchedObject, env, path.extend(i))
        })

    fun makePredicate(pattern: Pattern, matchedObject: VariableReference, path: PatternPath=PatternPath.EMPTY): CoreExpression = when (pattern) {
        is Pattern.Wildcard     -> CoreExpression.TRUE
        is Pattern.Variable     -> CoreExpression.TRUE
        is Pattern.Literal      -> CoreEqualConstantExpression(pattern.value, CoreExtractExpression(matchedObject, path))
        is Pattern.Constructor  -> constructorPredicate(pattern, path, matchedObject)
    }

    private fun constructorPredicate(pattern: Pattern.Constructor, path: PatternPath, matchedObject: VariableReference): CoreExpression {
        val exps = ArrayList<CoreExpression>()
        exps.add(matchesConstructor(path, pattern.name, matchedObject))

        pattern.args.forEachIndexed { i, p ->
            exps.add(makePredicate(p, matchedObject, path.extend(i)))
        }

        return CoreExpression.and(exps)
    }

    private fun matchesConstructor(path: PatternPath, name: String, matchedObject: VariableReference): CoreExpression =
        CoreEqualConstantExpression(name, CoreExtractTagExpression(matchedObject, path))
}
